tunnel/intercept/ebpf/ebpf_linux.go

```diff
@@ -0,0 +1,520 @@
+	if self.udpLn != nil {
+		return (*UDPIPPortAddr)(self.udpLn.LocalAddr().(*net.UDPAddr))
+	}
+
+	logrus.Errorf("invalid state: no udp listener for tproxy[%s]", self.service.Name)
+	return nil
+}
+
+func (self *eBpf) Intercept(resolver dns.Resolver, tracker intercept.AddressTracker) error {
+	service := self.service
+	if service.InterceptV1Config == nil {
+		return errors.Errorf("no client configuration for service %v", service.Name)
+	}
+
+	config := service.InterceptV1Config
+	logrus.Debugf("service %v using intercept.v1", service.Name)
+	var ports []IPPortAddr
+	for _, p := range config.Protocols {
+		if p == "tcp" {
+			logrus.Debugf("service %v intercepting tcp", service.Name)
+			ports = append(ports, self.tcpPort())
+		} else if p == "udp" {
+			logrus.Debugf("service %v intercepting udp", service.Name)
+			ports = append(ports, self.udpPort())
+		}
+	}
+
+	return self.intercept(service, resolver, ports, tracker)
+}
+
+func (self *eBpf) Apply(addr *intercept.InterceptAddress) {
+	logrus.Debugf("for service %v, intercepting proto: %v, cidr: %v, ports: %v:%v", self.service.Name, addr.Proto(), addr.IpNet(), addr.LowPort(), addr.HighPort())
+
+	var port IPPortAddr
+	switch addr.Proto() {
+	case "tcp":
+		port = self.tcpPort()
+	case "udp":
+		port = self.udpPort()
+	default:
+		logrus.Errorf("unknown proto[%s] for tproxy[%s]", addr.Proto(), self.service.Name)
+		return
+	}
+	if err := self.addInterceptAddr(addr, self.service, port, self.tracker); err != nil {
+		logrus.Debugf("failed for service %v, intercepting proto: %v, cidr: %v, ports: %v:%v", self.service.Name, addr.Proto(), addr.IpNet(), addr.LowPort(), addr.HighPort())
+
+		// do we undo the previous succesful ones?
+		// only fail at end and return all that failed?
+	}
+}
+
+func (self *eBpf) intercept(service *entities.Service, resolver dns.Resolver, ports []IPPortAddr, tracker intercept.AddressTracker) error {
+	var protocols []string
+	for _, p := range ports {
+		protocols = append(protocols, p.GetProtocol())
+	}
+
+	err := intercept.GetInterceptAddresses(service, protocols, resolver, self)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (self *eBpf) addInterceptAddr(interceptAddr *intercept.InterceptAddress, service *entities.Service, port IPPortAddr, tracker intercept.AddressTracker) error {
+	ipNet := interceptAddr.IpNet()
+	if err := router.AddLocalAddress(ipNet, "lo"); err != nil {
+		return errors.Wrapf(err, "failed to add local route %v", ipNet)
+	}
+	tracker.AddAddress(ipNet.String())
+	self.addresses = append(self.addresses, interceptAddr)
+
+	interceptAddr.TproxySpec = []string{
+		"-m", "comment", "--comment", service.Name,
+		"-d", ipNet.String(),
+		"-p", interceptAddr.Proto(),
+		"--dport", fmt.Sprintf("%v:%v", interceptAddr.LowPort(), interceptAddr.HighPort()),
+		"-j", "TPROXY",
+		"--tproxy-mark", "0x1/0x1",
+		fmt.Sprintf("--on-ip=%s", port.GetIP().String()),
+		fmt.Sprintf("--on-port=%d", port.GetPort()),
+	}
```

Comment:
Probably would want to consolidate this logging into a single statement. 
`pfxlog.Logger().WithField("dst_ip", ipNet).WithField("protocol", interceptAddr.Proto()).<rest of fields>.Info("setting up udp intercept)`

Reply:
Hi Paul I consolidated logging statements as recommended and also refactored intercept add/delete removing protocol conditional statements since it was redundant code.